zstackctl/zstackctl/ctl.py

```
@@ -4092,6 +4093,18 @@ def run(self, args):
         for file in [args.mevoco_installer, community_iso]:
             self.check_file_exist(file, UpgradeHACmd.host_post_info_list)
 
+        # pre-check all HA nodes before stopping any node
+        mevoco_dir = os.path.dirname(args.mevoco_installer)
+        mevoco_bin = os.path.basename(args.mevoco_installer)
+        for host_post_info in UpgradeHACmd.host_post_info_list:
+            command = "rm -rf /tmp/zstack_upgrade.lock && cd %s && bash %s -u -i --precheck" % (
+                mevoco_dir, mevoco_bin)
+            (status, output) = subprocess.getstatusoutput(
```

Ruff (0.15.9)

[error] Starting a process with a shell, possible injection detected

(S605)